02-intermediate-ruby/api-intro.md

```diff
@@ -1,7 +1,215 @@
+Content-Length: 21
+Connection: Keep-Alive
+
+# then a bunch of HTML or JSON
+```
+
+There are two important things here. At the top we see the _status code_, in this case `200 OK`. A 200 response indicates that the server was able to process the request without any problems.
+
+We also see "a bunch of HTML or JSON", more formally referred to as a _payload_. The payload is optional, and is usually only included in the response to a GET request.
+
+Again, we can see there's a bunch of other HTTP stuff in there that we're ignoring for right now.
+
+### Status Codes
+
+Status codes are an important part of the HTTP protocol - they let the client know how things went. Every response must have a status code attached.
+
+Each status code is a 3-digit number paired with an English word or phrase. The first digit of the number indicates what sort of status it is, and the last two tell the details. It's sort of like how the first two digits of a room number are the floor that room is on.
+
+Here are some common status codes:
+
+Code    | Text             | Meaning
+---     | ---              | ---
+**2XX** | **Success**      |
+200     | OK               | Standard success response, typically has an HTML payload (common for `GET` requests)
+201     | Created          | We made a new resource for you (common for `POST` requests)
+204     | No Content       | We did what you asked and it went fine, but there's no HTML to give you (common for `DELETE` requests)
+**3XX** | **Redirection**  |
+302     | Found            | Go look at this other URL
+**4XX** | **Client Error** |
+400     | Bad Request      | The data you sent doesn't make sense
+401     | Unauthorized     | You need to log in to do that
+403     | Forbidden        | The person you're logged in as doesn't have permission to do that
+404     | Not Found        | The resource you requested doesn't exist
+**5XX** | **Server Error** |
+500     | Internal Server Error | There's a bug in this Rails project. As engineers, we see this one a lot!
+
+**If you're writing a client program, you should make sure to check the response code for every request you send!**
+
+## Summary
+
+- The _client/server_ architecture is very common on the internet
+  - The server is a dedicated computer that keeps track of some information
+  - The client is an application (command-line program or the browser) that wants that information
+  - There are typically many clients for every server
+- Whenever computers communicate with each other, they use a _protocol_
+  - All the work we do in this class will involve the HTTP protocol
+- There are two main types of resource on the web:
+  - _Websites_ are designed to be used by humans
+  - _APIs_ are designed to be used by other machines
+  - Browsers can consume both
+- A key pattern on the web is the _request/response cycle_
+  - Client makes a request, including HTTP verb, path, optional payload
+  - Server sends a response, including HTTP status code, optional payload
+
```

Comment:
You mentioned this before, but this lesson is pretty long/has a lot of important new information. It might be nice to have a series of high-level comprehension questions at the end here (before or after the Summary section. I like after the Summary section because it has a lot of the answers hehe).

For example, if I were giving this lesson/writing the questions, I would probably ask students to answer this with the people next to them:

- Describe the relationship between client and server-- if we write a Ruby CLI that makes a request to an API, which is the client and which is the server?
- What data type did the API `dog.ceo` (from the example above) give back (looking for JSON)?
- What are the HTTP Verbs?

If you want to pursue this, also... maybe just a series of Summary/Details tags? :D Then they would have the answers if we ran out of time in lecture.